wp-graphql/wp-graphql
# feat: implement SemVer-compliant update checker

What does this implement/fix? Explain your changes.
---------------------------------------------------
This PR implements Semantic Version checking when updating WPGraphQL to a newer version.

### How it works:

- A new `Admin\Updates\UpdateChecker` class has been introduced to check whether an update is likely "non-breaking" (per Semantic Versioning). This class is used both to prevent auto-updates and to display update information before a manual update is applied.

- The `UpdateChecker` class also checks if there are any _explicitly incompatible_ or _possibly untested_ WPGraphQL extensions that should be checked/updated before (auto)updating.

  - New Plugin headers for extensions:
  
     - `Requires WPGraphQL: {x.y.z}` sets a minimum plugin version.
     - `WPGraphQL tested up to: {x.y.z}` sets the last _known_ compatible version.

    Plugins that use the `Requires Plugins: wp-graphql` or use `WPGraphQL` in the plugin name or description are detected with an `Unknown` version compatibility.

- The `PluginScreenLoader` and `UpdateScreenLoader` classes inject the notices/modals into the admin screens to warn and confirm about untested extensions that are potentially breaking.

  - A new `packages/updates` package contains the `.scss` to style our notices/modals. Additionally, `webpack-remove-empty-scripts` has been added to our NPM deps to prevent the generation of an empty `.js` file.

### New filters:

-  `wpgraphql_enable_major_autoupdates` - allows for major versions to be auto-updateable. Defaults `false`.
-  `wpgraphql_untested_release_type` - determines which SemVer release type should be considered potentially "breaking". Defaults `major`.
-  `graphql_get_dependents` and `graphql_get_possible_dependents` filter the list of (possible) WPGraphQL extensions.

### Todo
- [x] Implement APIs
- [x] Implement Autoupdate Prevention
- [x] Prevent activation ( + notice ) of plugins that don't meet the `Requires WPGraphQL` version
- [x] Inject Upgrade notice on Plugin Screen
- [x] Implement warning modals on Manual Update.
- [x] Attribute prior art.
- [ ] (Bonus) Show Upgrade Notices from `readme.txt`
- [ ] (Bonus) Explore reuse in semver-compatible extensions.

Does this close any currently open issues?
------------------------------------------


Closes #2543

Testing Instructions
-------------------
1. Install a WPGraphQL Extension. (you can copy the mock extensions from `test/data/plugins` into `wp-content/plugins` )
2. Mock an available update:

   ```php
   add_action( 'init', static function() {
	$update_plugins = get_site_transient( 'update_plugins' );

	if ( ! $update_plugins ) {
		return;
	}

	// Prime.
	if ( empty( $update_plugins->response ) ) {
		$update_plugins->response = [];
	}

	if ( isset( $update_plugins->no_update['wp-graphql/wp-graphql.php'] ) ) {
		$update_plugins->response['wp-graphql/wp-graphql.php'] = $update_plugins->no_update['wp-graphql/wp-graphql.php'];

		unset( $update_plugins->no_update['wp-graphql/wp-graphql.php'] );
	}

	// Stub new version.
	if ( isset( $update_plugins->response['wp-graphql/wp-graphql.php'] ) ) {
		// Ensure array.
		$update_plugins->response['wp-graphql/wp-graphql.php']->new_version = '4.0.0';
	}

	set_site_transient( 'update_plugins', $update_plugins );
   } );
   ```
   
2. **Updates Screen**:
  a. Visit `/wp-admin/update-core.php` and refresh the screen a few times* until the update is available.
  b. Select `WPGraphQL` from the list of plugins to view the modal.
  c. Click `Cancel` to see WPGraphQL get unchecked.
  
3. **Plugins Screen**
  a. Visit `wp-admin/plugins.php` and refresh the screen a few times* until the update is available.
  b. Click the inline `Update now` link to view the modal.

> [!WARNING]
> **DO NOT** try to update WPGraphQL to the fictitious version. (e.g. clicking the `Update now` button in the *modal*).
> It will delete the existing `wp-graphql` plugin, including the sym-linked reference to the repo.

Any relevant logs, error output, GraphiQL screenshots, etc?
-------------------------------------


### Plugin Screen notice:
![image](https://github.com/user-attachments/assets/d342ee84-5e9b-453f-8cd2-30a2ec1faedd)

### Plugin Screen modal:
![image](https://github.com/user-attachments/assets/24e5f919-a8a3-4d0d-9a34-6e98c4ecb104)

### Update Screen modal:
![image](https://github.com/user-attachments/assets/bf5eb1be-4e48-49ca-9b2d-9ce8243312bb)

### Deactivated Plugin notice
![image](https://github.com/user-attachments/assets/f766376b-d011-4575-9156-9c3ee08697fc)

Any other comments?
-------------------

- Displaying Upgrade Notes:
   Possibly beyond the scope, but if as part of our Upgrade Policy (#3218) we include Upgrade Notes in the dist version of the plugin somewhere, we can display them in our notices. Prior art: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/admin/plugin-updates/class-wc-plugins-screen-updates.php#L80
   
- SemVer-compliant Extension Updates
  A personal goal is to make the API reusable enough that any WPGraphQL extension could easily tap into it to make their own updates SemVer compliant. While that's _definitely_ beyond the scope, I will be auditing the API/method signatures so we can support this with non-breaking iteration.

Where has this been tested?
---------------------------
**Operating System:** Ubuntu 24.04 ( wsl2 + ddev + php8.2.26)

**WordPress Version:** 6.7.1
